### PR TITLE
Expose opening directories just for traversal (O_PATH)

### LINKED ISSUE #3741: Dir.cwd doesn't support iteration
Calling `std.fs.Dir.cwd().iterate().next()` fails with `EBADF` on Linux: `getdents` and `getdents64` don't support `AT_FDCWD` (and can't, since they store iteration state in the file descriptor). I'm not certain of the behavior on BSDs and Windows, but I believe that the same problem shows up. Note that iterating over `Dir.open(".")` works fine.

I'd be happy to fix this, but I'm not certain of the correct approach. I see three possible answers:
- Normalize and document this behavior. This would make a distinction between files opened for reading (which can be iterated over) and files opened purely for sub-file operations (i.e. with `O_PATH` or `AT_FDCWD`). This would presumably also involve splitting `openDir` into two functions for opening in the two different ways, and possibly a `reopenForReading` function that is a shorthand for `self.openDirRead(".")`.
- Detect `Dir.cwd()` in `iterate` and replace the file descriptor with a newly opened handle to `.`. This would make `iterate` fallible where it currently isn't.
- Detect `Dir.cwd()` in `next` and replace the file descriptor with a newly opened handle to `.`. This would add at least `ProcessFdQuotaExceeded` and `SystemFdQuotaExceeded` and probably `SystemResources` to the `Iterator.Error` set.

The latter two answers still don't make `Dir.cwd()` act like `Dir.open(".")` with respect to iteration, but this is also an inconsistency between systems. On Linux, since the iteration state is stored by the operating system in the directory file descriptor, calling `iterate` multiple times won't actually produce distinct new iterators, but rather iterators that are both grabbing entries off of the same list and interfering with each other. On the BSDs, the iteration state is an offset stored user-side, and so different iterators will actually be independent. On Windows, I think you get BSD-style behavior if the iterations are done one after the other (due to the `RestartScan` argument), but an even messier version of the Linux behavior if the iterators are used concurrently. The special-casing of `Dir.cwd()` would produce BSD-like behavior just for that one special case, even on Linux.

I personally would favor the first approach, if for no other reason than liking exposing the `O_PATH` optimization.

### PR DESCRIPTION
**Pull request by gereeter**
Closes #3741. Further bikeshedding is welcome.

### CODE REVIEW

File: lib/std/fs.zig [hunk 1]
```diff
@@ -774,25 +774,68 @@ pub const Dir = struct {
...
             const sub_path_w = try os.windows.cStrToPrefixedFileW(sub_path_c);
-            return self.openDirW(&sub_path_w);
+            return self.openDirListW(&sub_path_w);
+        } else {
+            return self.openDirFlagsC(sub_path_c, os.O_RDONLY | os.O_DIRECTORY | os.O_CLOEXEC);
         }
+    }
 
-        const flags = os.O_RDONLY | os.O_DIRECTORY | os.O_CLOEXEC;
+    fn openDirFlagsC(self: Dir, sub_path_c: [*]const u8, flags: u32) OpenError!Dir {
```
Comment (daurnimator) [inline lib/std/fs.zig]: I'd either `or` `O_DIRECTORY` into flags here; or `assert` that it is set.

**Comment by daurnimator** [comment]
Needs rebasing/updating now that sentinel types exist.

**Comment by gereeter** [comment]
Rebased and moved `O_DIRECTORY`. I initially kept that flag out of `openDirFlagsC` because it was purely for preventing duplication of code and it felt weird to partially but not completely deduplicate the flags, but you are right that it fits the meaning of the function better.

**Comment by gereeter** [comment]
I'm pretty sure that the test failure is the deadlock mentioned in #3751. I'd have liked to have a proper Windows test run, though.

**Comment by andrewrk** [comment]
The code looks good. I'll run it locally on Windows when merging.

What do you think about a slight modification (that changes docs only), making `std.Dir.openDir` defined to open for *both* traversal and listing, openDirList for only listing, and openDirTraverse for only traversal?

**Comment by gereeter** [comment]
> I'll run it locally on Windows when merging.

Great! (It did succeed on CI before the rebase, so I do full expect it to pass. But better safe than sorry, especially with confusingly documented semantics.)

> What do you think about a slight modification (that changes docs only), making `std.Dir.openDir` defined to open for *both* traversal and listing, `openDirList` for only listing, and `openDirTraverse` for only traversal?

(Technically, this affects more than docs since it would mean taking traversal away from `openDirList` if I understand correctly.) I considered something like this, but:
- Keeping `openDir` around seems like giving the user a suboptimal (too permissive and possibly unnecessarily expensive) default instead of forcing them to consider what the correct choice is, which seems against Zig philosophy.
- At that point you have 3 functions each with 3 variants, and I'd think it would make more sense to curb that explosion by passing a set of flags like the underlying OS APIs do (which I wouldn't be opposed to, but see the next point).
- You basically never want to list the contents of a directory without also traversing it. The only use-cases I can think of are reimplementing `ls` and glob expansion, and both of those support recursive enumeration that would require traversal anyway. (And most cases where people do in practice list separately from traversal they are introducing ToCToU bugs since they open the directory twice and it may have changed.)

However, if you want me to move to a flag-based API or something like that, I can.

**Comment by andrewrk** [comment]
Alright, I'm convinced to merge it as is. Thanks for taking the time to explain this reasoning.

**Pull request closed without merging**